### DISCUSSION TITLE
map to function to run multiple actions i.e. zen mode

### DISCUSSION THREAD

**Discussion by jimcornmell** [kovidgoyal/kitty#3799] [Q&A] [answered]
I'm interested to create a keyboard mapping to go to a zen mode, which makes sets full screen and increases the font size.  Something like the following in the kitty config file.   Is this possible?

     map kitty_mod+z  toggle_zenmode

     function toggle_zenmode
           if inzenmode
                 change_font_size all -4.0
                 inzenmode = false
           else
                 change_font_size all +4.0
                 inzenmode = true
           endif

           toggle_fullscreen
    endfunction

**Comment by kovidgoyal** [comment]
See the very first example here:
https://sw.kovidgoyal.net/kitty/kittens/custom.html

**Reply by jimcornmell** [reply]
This seems to work:

    def main(args):
         pass


    from kittens.tui.handler import result_handler


    @result_handler(no_ui=True)
    def handle_result(args, answer, target_window_id, boss):
        boss.change_font_size("all", "+", 4.0)
        boss.toggle_fullscreen()

**Comment by jimcornmell** [comment]
One last question (promise)  How can you tell you are in fullscreen mode?  I can see in the C there is "is_window_fullscreen"  but cannot see the equivalent in the python API.

**Comment by kovidgoyal** [comment]
On Mon, Jul 05, 2021 at 03:38:58AM -0700, jimcornmell wrote:
> One last question (promise)  How can you tell you are in fullscreen mode?  I can see in the C there is "is_window_fullscreen"  but cannot see the equivalent in the python API.

There isn't. Fullscreen can be triggerred in many ways, from outside of
kitty as well. And there are different types of fullscreen on different
platforms.

**Reply by jimcornmell** [reply]
Hi, In that case whats the easiest way to persist a value between instantiations of the kitten without resorting to databases or system variables?

**Comment by kovidgoyal** [comment]
On Mon, Jul 05, 2021 at 05:20:15AM -0700, jimcornmell wrote:
> Hi, In that case whats the easiest way to persist a value between instantiations of the kitten without resorting to databases or system variables?

The UI part of the kitten is invoked in the same process, just use a
global variable.

**Reply by jimcornmell** [reply]
Does not appear to work I have this in kitty.conf:

    map kitty_mod+z kitten zenmode.py

And this in zenmode.py:

    from kittens.tui.handler import result_handler

    def main(args):
        pass

    @result_handler(no_ui=True)
    def handle_result(args, answer, target_window_id, boss):
        global inZenMode

        try:
            inZenMode
        except:
            inZenMode = False

        if inZenMode:
            boss.change_font_size("all", "-", 5.0)
            inZenMode = False
        else:
            boss.change_font_size("all", "+", 5.0)
            inZenMode = True

        boss.toggle_fullscreen()

**Reply by kovidgoyal** [reply]
You need an actual inZenMode outside the function. the global keyword doesnt make a variable global only declares that the local variable does not shadow the global one.

**Reply by jimcornmell** [reply]
Hi,  Sorry I'm not a python programmer, but that does not appear to work either:

    from kittens.tui.handler import result_handler

    def main(args):
        pass

    try:
        inZenMode
    except:
        inZenMode = False

    @result_handler(no_ui=True)
    def handle_result(args, answer, target_window_id, boss):
        global inZenMode

        try:
            inZenMode
        except:
            inZenMode = False

        if inZenMode:
            boss.change_font_size("all", "-", 5.0)
            inZenMode = False
        else:
            boss.change_font_size("all", "+", 5.0)
            inZenMode = True

        boss.toggle_fullscreen()

**Comment by kovidgoyal** [comment, marked as answer]
```py
from kittens.tui.handler import result_handler


def main(args):
    pass


@result_handler(no_ui=True)
def handle_result(args, answer, target_window_id, boss):
    in_zen_mode = getattr(boss, 'in_zen_mode', False)
    boss.change_font_size("all", "-" if in_zen_mode else '+', 5.0)
    boss.toggle_fullscreen()
    boss.in_zen_mode = not in_zen_mode
```

**Reply by jimcornmell** [reply]
Perfect!   Many thanks, I've learnt a lot.
